src/map.c

```diff
@@ -3,752 +3,630 @@
  * "LICENSE" for information on usage and redistribution of this file.
  */
 
+/*
+ * This map implementation has undergone extensive modifications, heavily
+ * relying on the rb.h header file from jemalloc. See
+ * https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/rb.h
+ * . The original rb.h file served as the foundation and source of inspiration
+ * for adapting and tailoring it specifically for this map implementation.
+ * Therefore, credit and sincere thanks are extended to jemalloc for their
+ * invaluable work.
+ */
+
 #include <assert.h>
-#include <stdint.h>
-#include <stdlib.h>
-#include <string.h>
+#include <memory.h>
 
 #include "map.h"
 
-struct map_internal {
-    struct map_node *head;
-
-    /* Properties */
-    size_t key_size, element_size, size;
+#define RB_MAX_DEPTH (sizeof(void *) << 4)
 
-    map_iter_t it_end, it_most, it_least;
-
-    int (*comparator)(const void *, const void *);
-};
-
-typedef enum { RB_RED = 0, RB_BLACK } map_color_t;
-
-/*
- * Get parent of node
- * @node: pointer to the rb node
- *
- * Return: rb parent node of @node
- */
-static inline map_node_t *rb_parent(const map_node_t *node)
+/* Left accessors */
+static inline map_node_t *rbtn_left_get(map_node_t *node)
 {
-    return (map_node_t *) (node->parent_color & ~1LU);
+    return node->link.left;
 }
 
-/*
- * Get color of node
- * @node: pointer to the rb node
- *
- * Return: color of @node
- */
-static inline map_color_t rb_color(const map_node_t *node)
+static inline void rbtn_left_set(map_node_t *node, map_node_t *left)
 {
-    return (map_color_t) (node->parent_color & 1LU);
+    node->link.left = left;
 }
 
-/*
- * Set parent of node
- * @node: pointer to the rb node
- * @parent: pointer to the new parent node
- */
-static inline void rb_set_parent(map_node_t *node, map_node_t *parent)
+/* Right accessors */
+static inline map_node_t *rbtn_right_get(map_node_t *node)
 {
-    node->parent_color = (unsigned long) parent | (node->parent_color & 1LU);
+    return (map_node_t *) (((size_t) node->link.right_red) & ~3);
 }
 
-/*
- * Set color of node
- * @node: pointer to the rb node
- * @color: new color of the node
- */
-static inline void rb_set_color(map_node_t *node, map_color_t color)
+static inline void rbtn_right_set(map_node_t *node, map_node_t *right)
 {
-    node->parent_color = (node->parent_color & ~1LU) | color;
+    node->link.right_red =
+        (map_node_t *) (((size_t) right) |
+                        (((size_t) node->link.right_red) & 1));
 }
 
-/*
- * Set parent and color of node
- * @node: pointer to the rb node
- * @parent: pointer to the new parent node
- * @color: new color of the node
- */
-static inline void rb_set_parent_color(map_node_t *node,
-                                       map_node_t *parent,
-                                       map_color_t color)
+/* Color accessors */
+static inline bool rbtn_red_get(map_node_t *node)
```

Comment:
Can you make the naming scheme more consistent? That is, something like `rbnode_get_red`.